Add CheckboxListTile checkColor

christian-muertz opened the pull request:
## Description

* Add missing checkColor property to the `CheckboxListTile` widget.

## Related Issues

* Fixes #37606

## Tests

I added the following tests:

* Some basic tests to check that the color is passed correctly.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.
- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

christian-muertz committed:
Add checkColor to CheckboxListTile

Add missing checkColor property to the CheckboxListTile widget. Fixes #37606.

Commit:
Fix typo in Checkbox checkColor documentation

christian-muertz commented:
@HansMuller

shihaohong commented:
I would add a test to ensure that the passed in checkColor property is properly applied (as well as a counter-test to prove that the default color is what it claims to be). 

There is currently a control test for `CheckboxListTile` in `flutter/packages/test/material/control_list_tile_test.dart`, so that could be a good place to add a test. Otherwise, creating a new file for `checkbox_list_tile_test.dart` and moving any existing tests over might be a better option.

shihaohong commented:
Also, `flutter/packages/test/material/checkbox_test.dart` has great pointers for how you would construct your tests!

christian-muertz commented:
Argg think i found a bug in the canvas recording...

Please confirm that this is not the expected behaviour:
 
` expect(getCheckboxListTileRenderer(), paints..rrect(color: someExpectedColor);`

fails when the paint used to paint the rect is changed afterwards. This is the case for example in `_RenderCheckbox.paint`:

```
 final RRect outer = _outerRectAt(origin, 1.0);
 final Paint paint = Paint() ..color = _colorAt(1.0);
 canvas.drawRRect(outer, paint);

 _initStrokePaint(paint);
```   
with 
```
  void _initStrokePaint(Paint paint) {
    paint
      ..color = checkColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = _kStrokeWidth;
  }
```

So the expect line fails because the `_TestRecordingCanvasPatternMatcher` says that the color of the rect is the `checkColor` and not the color `_colorAt(1.0)`

christian-muertz commented:
Ok the `_DrawCommandPaintPredicate` argument verifier just uses the paint object to verify the color. If the paint color changes after the draw call the test fails.

I would suggest to use a new `Paint` for drawing the check. This would allow me to test if the `activeColor` is applied correctly. Another solution would be to change the `_DrawCommandPaintPredicate` to use the actual color used to paint instead of just the color of the `Paint` object passed.

Commit:
Create second paint for drawing check and dash

Create a second Paint object to draw the check and dash of the Checkbox widget.

Commit:
Add basic tests for CheckboxListTile widget

christian-muertz commented:
`CheckboxListTile.activeColor` says:
```
Defaults to accent color of the current [Theme].
```
Which only applies for the `ListTile` `selectedColor` but not for the checkbox.

`Checkbox.activeColor` says:
```
Defaults to [ThemeData.toggleableActiveColor].
```

Which one is correct??

Commit:
Add activeColor test for Checkbox

shihaohong commented on packages/flutter/lib/src/material/checkbox.dart:
```diff
@@ -281,8 +281,8 @@ class _RenderCheckbox extends RenderToggleable {
   }
 
   // White stroke used to paint the check and dash.
```
I'm not sure if white stroke is accurate

shihaohong commented on packages/flutter/test/material/checkbox_list_tile_test.dart:
```diff
@@ -0,0 +1,85 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
```suggestion
// Copyright 2019 The Chromium Authors. All rights reserved.
```

shihaohong reviewed (approved):
LGTM! Thank you for adding comprehensive tests and isolating checkbox_list_tile_test.dart into its own file.

Commit:
Fix CheckboxListTile test file copyright

Co-Authored-By: Shi-Hao Hong <shihaohong@google.com>
